### ISSUE TITLE
[Question] Sampling from distribution

### ISSUE THREAD

**Issue by suargi** [DLR-RM/stable-baselines3#2096]
### ❓ Question

This is a general question regarding the implementation in SB3.

In the normal pytorch behaviour, when using `.rsample()` to obtain actions from a Gaussian distribution, the gradients with respect to the mean parameters of the policy network are zero. This does not happen when using `.sample()`. 

In SB3 the actions are sampled using `.rsample()`, but the gradients are indeed different from zero. Before I delve into the code, is there some implementation trick in SB3 that solves this _problem_?

Here is a piece of code to test it. Thanks!

``` python
import torch as th
import torch.nn as nn
from torch.autograd import grad
from torch.distributions import Normal

th.manual_seed(0)

class Policy(nn.Module):
    def __init__(self):
        super(Policy, self).__init__()

        self.mean = nn.Linear(10, 2)

        self.log_std = nn.Parameter(th.Tensor(2))
        nn.init.constant_(self.log_std, 0)

    def density(self, state):
        actions_mean = self.mean(state)
        scale = self.log_std.exp().expand_as(actions_mean)
        return Normal(loc=actions_mean, scale=scale)

    def forward(self, state):
        density = self.density(state=state)
        actions = density.rsample()
        return actions, density.log_prob(actions).sum(dim=1)

model = GaussianMLPPolicy()

x = th.rand((5, 10))
_, log_prob = model(x)

loss = log_prob.mean()

gradients = grad(loss,
                 list(model.parameters()))

# gradients w.r.t. mean are zero with .rsample

print(gradients)
```

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [x] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [x] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
https://stackoverflow.com/questions/60533150/what-is-the-difference-between-sample-and-rsample ?

**Comment by suargi** [comment]
Thank you for your answer. I have edited the original question as it was totally missing the point? :)

**Comment by suargi** [comment]
The buffer stores observations, actions, and other relevant variables collected during trajectory sampling. When recomputing `log_probs`—such as in A2C—SB3 does **not** use the `forward` method with the collected observations. Instead, it calls `evaluate_actions`, which takes both the stored observations and actions as input. This ensures that gradients are correctly computed for the mean network.  

This approach is necessary because, during trajectory collection, actions and `log_probs` are computed within a `torch.no_grad()` block. As a result, `log_probs` are **not** stored in the computation graph. Simply calling `forward` again would resample actions from the normal distribution, leading to different values. The `evaluate_actions` method avoids this issue by computing `log_probs` without resampling, ensuring consistency.  

Interestingly, using `.sample()` with the `forward` method (without `torch.no_grad()`) produces the same gradients as using `.rsample()` with `evaluate_actions`. But why does `.sample()` work in this case?  

While `.sample()` does not allow gradients to flow through the sampled actions, the policy gradient loss does **not** depend on the actions themselves—only on `log_prob`, which still depends on the mean parameters. This allows us to obtain **nonzero gradients**. In contrast, the reparameterization trick (`.rsample()`) leads to zero gradients in expectation due to how it expresses sampled actions in terms of the mean and noise, as demonstrated in the first script.

I hope it is clearer.
I think SB3 could use rsample or sample.
